apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,324 @@
+    if auth_idx then
+        local protocol_idx = str_find(url, "://")
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local arr = ngx_re.split(user_and_password, ":")
+        if #arr == 2 then
+            username = arr[1]
+            password = arr[2]
+        end
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+    end
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    return url, username, password
+end
+
+
+local function request(request_uri, path, body, method, basic_auth)
+    local url = request_uri .. path
+    log.info("request url:", url)
+    local headers = core.table.new(0, 0)
+    headers['Connection'] = 'Keep-Alive'
+    headers['Accept'] = 'application/json'
+
+    if basic_auth then
+        headers['Authorization'] = basic_auth
+    end
+
+    if body and 'table' == type(body) then
+        local err
+        body, err = core.json.encode(body)
+        if not body then
+            return nil, 'invalid body : ' .. err
+        end
+        -- log.warn(method, url, body)
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.discovery.nacos.timeout
+    local connect_timeout = timeout.connect
+    local send_timeout = timeout.send
+    local read_timeout = timeout.read
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+        ", read_timeout:", read_timeout, ".")
+    httpc:set_timeouts(connect_timeout, send_timeout, read_timeout)
+    local res, err = httpc:request_uri(url, {
+        method = method,
+        headers = headers,
+        body = body,
+        ssl_verify = false,
+    })
+    if not res then
+        return nil, err
+    end
+
+    if not res.body or res.status ~= 200 then
+        return nil, "status = " .. res.status
+    end
+
+    local json_str = res.body
+    local data, err = core.json.decode(json_str)
+    if not data then
+        return nil, err
+    end
+    return data
+end
+
+
+local function get_url(request_uri, path)
+    return request(request_uri, path, nil, "GET", nil)
+end
+
+
+local function post_url(request_uri, path, body)
+    return request(request_uri, path, body, "POST", nil)
+end
+
+
+
+local function get_page_service(infos, base_uri, token_param, page_num)
+    local path = str_format(service_list_path, page_num, page_size) .. token_param
+    local data, err = get_url(base_uri, path)
+    if err then
+        return data, err, path
+    end
+
+    for _, service_name in ipairs(data.doms) do
+        core.table.insert(infos, service_name)
+    end
+    return data, err, path
+end
+
+
+local function iter_and_add_service_info(infos, base_uri, token_param)
```

Comment:
Emm... I mean to iterate the upstream configurations and collect the service list from those that have `discovery_type` set to `nacos`. There is no need to ask nacos for the service list, we already know it via the upstream configuration.

Reply:
You can take a look at https://github.com/apache/apisix/blob/49e4fd307c65dc125e62987b3634a9e7358179bf/apisix/control/v1.lua#L115-L122
for how we iterate the upstream configurations and collect the health check info.
It is similar to collect the service_name.

Reply:
If import these in nacos.lua:
local upstream_mod = require("apisix.upstream")
local get_upstreams = upstream_mod.upstreams

When run test file,will throw error:
attempt to index local 'upstream_mod' (a number value)

It's not right import?

Reply:
I have mentioned it before: https://github.com/apache/apisix/pull/3820#discussion_r599481973

> You might need some lazy load trick to break the dependent cycle.

Reply:
Could you give me some demo code?

Reply:
Don't worry, I can update it for you once you have solved the other issues.